### ISSUE TITLE
Add a command to change the language of the active file

### ISSUE THREAD

**Issue by alexr00** [microsoft/vscode#125486]
This comes from @Kenobi-the-2nd in https://github.com/microsoft/vscode/issues/124907.

The general idea is that `workbench.action.editor.changeLanguageMode` is a useful command that users may want to use in tasks. And, as a command used in a task, it is helpful to be able to pass an argument for specifying the language so that the quick pick isn't needed.

Even if it isn't made an API command, allowing it to take an argument would make the command useful.

**Comment by bpasero** [comment]
We have an API for this, no?

https://github.com/microsoft/vscode/blob/e43c3957f64a9f231734818aa8f72e139dd1e01e/src/vs/vscode.d.ts#L10897-L10897

**Comment by alexr00** [comment]
The command can be used from a tasks.json task, but the API cannot (at least not without writing an extension that provides another command).

**Comment by bpasero** [comment]
But how much sense does it make to run it from a task? In the end it targets the currently active editor, isn't that a bit random?

**Comment by alexr00** [comment]
I will leave that question for @Kenobi-the-2nd.

**Comment by Kenobi-the-2nd** [comment]
@alexr00, thanks for opening this issue. 

@bpasero, I use this in two common scenarios. One is in a macro that I run to automatically format JSON records: 

        "format-json": [
            "workbench.action.editor.changeLanguageMode",
            "editor.action.selectAll",
            "extension.smartColumnIndenter.removeLineBreaks",
            "editor.action.formatDocument",
            "workbench.action.acceptSelectedQuickOpenItem"
        ],

I pull JSON records from various sources on a regular basis (such as from webservice calls, JSON line files, etc.) and would like a way to just copy this text into a new file and be able to format the JSON with one command/button so that I can inspect the record. Currently I have to change the language mode each time. 

The second scenario is when scripting out a SQL table definition. Much of the programming that I do is focused around database applications, so I am often needing to reference table definitions. I created a compound task with the pickString input option to do this (below). I would like to see syntax coloring on the SQL, so I change the language mode, but again, since the command doesn't accept a string argument, I have to change the language manually. 

        {
            "label": "Script Given Table as Create",
            "type": "shell",
            "dependsOrder": "sequence",
            "dependsOn": [
                "Script Table as Create",
                "Copy contents of table create script to clipboard",
                "Open new file",
                "Paste",
                "Go To File Start",
                "Change Language Mode",
                "Delete table create script"
            ],
            "problemMatcher": [],
            "isBackground": true,
            "presentation": {
                "echo": false,
                "reveal": "silent",
                "focus": false,
                "panel": "shared",
                "showReuseMessage": false,
                "clear": false
            },
            "runOptions": {
                "instanceLimit": 1,
                "runOn": "default"
            }
        },

**Comment by bpasero** [comment]
Thanks for the sample. I think this is still an extension candidate though because you can easily have an extension that defines a command that does something along the lines of:

```ts
const editor = window.activeTextEditor;
languages.setTextDocumentLanguage(editor.document, <language id>);
```

And the command could accept the language ID as parameter.

**Comment by Kenobi-the-2nd** [comment]
@bpasero I understand your reasoning, though I am still curious as to why the implementation for `workbench.action.editor.changeLanguageMode` does not accept a string as a parameter.

As far as extensions go that support this kind of behavior, I was able to find one (released in just the last month or so) that seems very flexible and exports a command that will allow passing a language id to change the language mode of the active editor. For those interested in this functionality, the extension is called [Commands](https://marketplace.visualstudio.com/items?itemName=usernamehw.commands). I was able to add the following, which solved my problem. 

From my settings.json:

```
"commands.commands": {
    "Change Language - SQL": {
        "command": "commands.setEditorLanguage",
        "args": "sql"
    }
}
```

From my tasks.json: 

```
{
    "label": "Change Language - SQL",
    "command": "${command:Change Language - SQL}",
}
```

**Comment by bpasero** [comment]
> why the implementation for `workbench.action.editor.changeLanguageMode` does not accept a string as a parameter

Because it is not an official API command like the other `vscode.*` commands. It is an internal workbench command that happens to be used by extensions, but if we make support with a parameter, I think it has to be an official API command.

**Comment by Kenobi-the-2nd** [comment]
Understood. @alexr00 and @bpasero, thank-you for your help on this issue. I appreciate your time.

**Comment by starball5** [comment]
Duplicate of #21826, which is incorrectly closed as a duplicate of #1800.

Duplicated by #135499.

Related on Stack Overflow:
- [How can I create a custom keybinding to change syntax highlighting to a specific language in Visual Studio Code?](https://stackoverflow.com/q/41639887/11107541)
- [In VS Code, how can I change to a specific language mode using a keybinding? \[duplicate\]](https://stackoverflow.com/q/56565851/11107541)
- [Visual Studio Code: keybinding to change document syntax highlighting to JSON \[duplicate\]](https://stackoverflow.com/q/53088885/11107541)
